// @chris.troutner/bch-js/src/bch-js.js
/*
  This is the primary library file for bch-js. This file combines all the other
  libraries in order to create the BCHJS class.
*/

// bch-api mainnet.
const DEFAULT_REST_API = "https://mainnet.bchjs.cash/v3/"

// local deps
const BitcoinCash = require("./bitcoincash")
const Crypto = require("./crypto")
const Util = require("./util")
const Blockchain = require("./blockchain")
const Control = require("./control")
const Generating = require("./generating")
const Mining = require("./mining")
const RawTransactions = require("./raw-transactions")
const Mnemonic = require("./mnemonic")
const Address = require("./address")
const HDNode = require("./hdnode")
const TransactionBuilder = require("./transaction-builder")
const ECPair = require("./ecpair")
const Script = require("./script")
const Price = require("./price")
const Socket = require("./socket")
const Wallet = require("./wallet")
const Schnorr = require("./schnorr")
const SLP = require("./slp/slp")

const InsightAddress = require("./insight/address")
const InsightBlock = require("./insight/block")
const InsightTransaction = require("./insight/transaction")

const Blockbook = require("./blockbook")
const Bitcore = require("./bitcore")
const OpenBazaar = require("./openbazaar")

class BCHJS {
  constructor(config) {
    // Try to retrieve the REST API URL from different sources.
    if (config && config.restURL && config.restURL !== "")
      this.restURL = config.restURL
    else if (process.env.RESTURL && process.env.RESTURL !== "")
      this.restURL = process.env.RESTURL
    else this.restURL = DEFAULT_REST_API

    // Populate the Insight API endpoints.
    this.Insight = {}
    this.Insight.Address = new InsightAddress(this.restURL)
    this.Insight.Block = new InsightBlock(this.restURL)
    this.Insight.Transaction = new InsightTransaction(this.restURL)

    // Populate Blockbook endpoints.
    this.Blockbook = new Blockbook(this.restURL)

    // Populate Bitcore endpoints
    this.Bitcore = new Bitcore(this.restURL)

    // Populate OpenBazaar endpoints
    this.OpenBazaar = new OpenBazaar(this.restURL)

    // Populate Full Node
    this.Control = new Control(this.restURL)
    this.Mining = new Mining(this.restURL)
    this.RawTransactions = new RawTransactions(this.restURL)

    // Populate utility functions
    this.Address = new Address(this.restURL)
    this.BitcoinCash = new BitcoinCash(this.Address)
    this.Blockchain = new Blockchain(this.restURL)
    this.Crypto = Crypto
    this.ECPair = ECPair
    this.ECPair.setAddress(this.Address)
    this.Generating = new Generating(this.restURL)
    this.HDNode = new HDNode(this.Address)
    this.Mnemonic = new Mnemonic(this.Address)
    this.Price = new Price()
    this.Script = new Script()
    this.TransactionBuilder = TransactionBuilder
    this.TransactionBuilder.setAddress(this.Address)
    this.Util = new Util(this.restURL)
    this.Socket = Socket
    this.Wallet = Wallet
    this.Schnorr = new Schnorr(this.restURL)

    this.SLP = new SLP(this.restURL)
    this.SLP.HDNode = this.HDNode
  }

  static BitboxShim() {
    return BitboxShim
  }
}

// Creat a shim so that all the endpoints match BITBOX SDK and this library
// can be used as a drop-in replacement for BITBOX.
class BitboxShim {
  constructor(config) {
    // Try to retrieve the REST API URL from different sources.
    if (config && config.restURL && config.restURL !== "")
      this.restURL = config.restURL
    else if (process.env.RESTURL && process.env.RESTURL !== "")
      this.restURL = process.env.RESTURL
    else this.restURL = DEFAULT_REST_API

    // Populate utility functions
    this.Address = new Address(this.restURL)
    this.BitcoinCash = new BitcoinCash(this.Address)
    this.Blockchain = new Blockchain(this.restURL)
    this.Crypto = Crypto
    this.ECPair = ECPair
    this.ECPair.setAddress(this.Address)
    this.Generating = new Generating(this.restURL)
    this.HDNode = new HDNode(this.Address)
    this.Mnemonic = new Mnemonic(this.Address)
    this.Price = new Price()
    this.Script = new Script()
    this.TransactionBuilder = TransactionBuilder
    this.TransactionBuilder.setAddress(this.Address)
    this.Util = new Util(this.restURL)
    this.Socket = Socket
    this.Wallet = Wallet
    this.Schnorr = new Schnorr(this.restURL)

    // Populate the SLP endpoints.
    this.SLP = new SLP(this.restURL)
    this.Address = this.SLP.Address
    // this.Address.toSLPAddress = this.SLP.Address.toSLPAddress
    // this.Address.isSLPAddress = this.SLP.Address.isSLPAddress
    // this.Address.toCashAddress = this.SLP.Address.toCashAddress
    // this.Address.toLegacyAddress = this.SLP.Address.toLegacyAddress
    // this.Address.isLegacyAddress = this.SLP.Address.isLegacyAddress
    // this.Address.isCashAddress = this.SLP.Address.isCashAddress
    // this.Address.isSLPAddress = this.SLP.Address.isSLPAddress
    // this.Address.isMainnetAddress = this.SLP.Address.isMainnetAddress
    // this.Address.isTestnetAddress = this.SLP.Address.isTestnetAddress
    // this.Address.isP2PKHAddress = this.SLP.Address.isP2PKHAddress
    // this.Address.isP2SHAddress = this.SLP.Address.isP2SHAddress
    // this.Address._ensureValidAddress = this.SLP.Address._ensureValidAddress
    // this.Address.detectAddressFormat = this.SLP.Address.detectAddressFormat
    this.ECPair.toSLPAddress = this.SLP.ECPair.toSLPAddress
    // this.HDNode.toSLPAddress = this.SLP.HDNode.toSLPAddress
    this.Util.list = this.SLP.Utils.list
    this.Util.balancesForAddress = this.SLP.Utils.balancesForAddress
    this.Util.balancesForToken = this.SLP.Utils.balancesForToken
    this.Util.balance = this.SLP.Utils.balance
    this.Util.validateTxid = this.SLP.Utils.validateTxid
    this.Util.tokenStats = this.SLP.Utils.tokenStats
    this.Util.transactions = this.SLP.Utils.transactions
    this.Util.burnTotal = this.SLP.Utils.burnTotal
    this.Util.decodeOpReturn = this.SLP.Utils.decodeOpReturn
    this.Util.isTokenUtxo = this.SLP.Utils.isTokenUtxo
    this.Util.tokenUtxoDetails = this.SLP.Utils.tokenUtxoDetails
    this.Util.txDetails = this.SLP.Utils.txDetails

    // Populate the Insight API endpoints.
    this.Insight = {}
    this.Insight.Address = new InsightAddress(this.restURL)
    this.Address.details = this.Insight.Address.details
    this.Address.utxo = this.Insight.Address.utxo
    this.Address.unconfirmed = this.Insight.Address.unconfirmed
    this.Address.transactions = this.Insight.Address.transactions

    // Populate Full Node
    this.Control = new Control(this.restURL)
    this.Mining = new Mining(this.restURL)
    this.RawTransactions = new RawTransactions(this.restURL)

    this.Insight.Block = new InsightBlock(this.restURL)
    this.Block = this.Insight.Block

    this.Insight.Transaction = new InsightTransaction(this.restURL)
    this.Transaction = this.Insight.Transaction

    // Populate Blockbook endpoints.
    this.Blockbook = new Blockbook(this.restURL)

    // Populate Bitcore endpoints
    this.Bitcore = new Bitcore(this.restURL)

    // Populate OpenBazaar endpoints
    this.OpenBazaar = new OpenBazaar(this.restURL)
  }
}

module.exports = BCHJS
